### PR TITLE
Owner image

### PR DESCRIPTION
**Pull request by Elkrival**
Added profile image next to the chart creator's name.

Screenshot 2023-12-17 at 5 26 17 PM

### CODE REVIEW

File: views/tables/ChartsTable/index.jsx [hunk 1]
```diff
@@ -48,7 +45,18 @@ const ChartsTable = ({
   const cellRenderer = (chart, property) => {
     switch (property) {
       case 'owner':
-        return chart['chartOwner'].display_name
+        return (
+          <div>
+            <Avatar
+              alt={chart['chartOwner'].display_name[0]}
+              src={chart['chartOwner'].icon}
+              sx={{ width: 24, height: 24, float: 'left' }}
```
Comment (mikestone14) [inline views/tables/ChartsTable/index.jsx]: we shouldn't have to use floats now that we have tools like flexbox and css grid.

**Comment by JackVCurtis** [comment]
Can we add the code from this example https://pragmaticwebsecurity.com/articles/spasecurity/react-xss-part1.html to sanitize the `data:` urls ~~before~~ after passing them to the `Avatar` component?

```javascript
const SAFE_URL_PATTERN = /^(?:(?:https?|mailto|ftp|tel|file|sms):|[^&:/?#]*(?:[/?#]|$))/gi;

/** A pattern that matches safe data URLs. It only matches image, video, and audio types. */
const DATA_URL_PATTERN = /^data:(?:image\/(?:bmp|gif|jpeg|jpg|png|tiff|webp)|video\/(?:mpeg|mp4|ogg|webm)|audio\/(?:mp3|oga|ogg|opus));base64,[a-z0-9+\/]+=*$/i;

function _sanitizeUrl(url: string): string {
  url = String(url);
  if (url === "null" || url.length === 0 || url === "about:blank") return "about:blank";
  if (url.match(SAFE_URL_PATTERN) || url.match(DATA_URL_PATTERN)) return url;

  return `unsafe:${url}`;
}

export function sanitizeUrl(url = "about:blank"): string {
  return _sanitizeUrl(String(url).trim());
}
```

React has no built-in validation for `data:` urls, but Angular does and this code is based on the Angular source code. This is a fairly obvious cross-site scripting vulnerability, a hacker could actively inject code into the site for anyone who sees their profile image.

File: views/tables/ChartsTable/index.jsx (hunk 1 above)
Comment (Elkrival) [reply views/tables/ChartsTable/index.jsx]: [c5452ad](https://github.com/AMP-SCZ/dpdash/pull/530/commits/c5452adac507cb21aef66b1d754d6d34b1f7e517)

**Comment by Elkrival** [comment]
@JackVCurtis [c5452ad](https://github.com/AMP-SCZ/dpdash/pull/530/commits/c5452adac507cb21aef66b1d754d6d34b1f7e517) I've added the suggested code. I had to update helmet for development purposes, it was throwing a refused to load script from livereload, production should be unaffected.

File: views/models/ChartModel/index.js [hunk 2]
```diff
@@ -1,5 +1,15 @@
+import { SAFE_URL_PATTERN, DATA_URL_PATTERN } from '../../../constants'
+
 const ChartModel = {
   isOwnedByUser: (chart, user) => chart.owner === user.uid,
+  sanitiazeProfileImageUrl: (url = 'about:blank') => {
```
Comment (mikestone14) [inline views/models/ChartModel/index.js]: do we need to default this?

File: views/models/ChartModel/index.js [hunk 3]
```diff
@@ -1,5 +1,15 @@
...
+    url = String(url)
+    if (url === 'null' || url.length === 0 || url === 'about:blank')
+      return 'about:blank'
+    if (url.match(SAFE_URL_PATTERN) || url.match(DATA_URL_PATTERN)) return url
+
+    return `unsafe:${url}`
```
Comment (mikestone14) [inline views/models/ChartModel/index.js]: should this throw an error instead of returning an unsafe URL?

File: views/models/ChartModel/index.js [hunk 4]
```diff
@@ -1,5 +1,15 @@
...
+  },
```
Comment (mikestone14) [inline views/models/ChartModel/index.js]:
```suggestion
  sanitiazeProfileImageUrl: (url) => {
    url = String(url)

    if (url.match(SAFE_URL_PATTERN) || url.match(DATA_URL_PATTERN)) {
      return url
    }

    throw new Error('Unsafe URL string')
  },
```

File: views/models/ChartModel/index.js (hunk 2 above)
Comment (mikestone14) [inline views/models/ChartModel/index.js]: I think we should pass the chart for this, or create a new UrlModel with a `sanitizeUrl` method that we can pass a URL to (or both).

File: views/models/ChartModel/index.js (hunk 3 above)
Comment (JackVCurtis) [reply views/models/ChartModel/index.js]: I think returning something is actually better, but maybe we can drop in a big "X" or "Unsafe URL". We don't want the app to choke for everybody if one person sets a bad profile picture

File: views/models/ChartModel/index.js (hunk 2 above)
Comment (JackVCurtis) [reply views/models/ChartModel/index.js]: The default was added in the example code I gave, but that was more generic so I think here it's unnecessary.

File: views/models/ChartModel/index.js (hunk 3 above)
Comment (JackVCurtis) [reply views/models/ChartModel/index.js]: Example return value: `data:,Unsafe%20Image`

File: views/models/ChartModel/index.js (hunk 2 above)
Comment (Elkrival) [reply views/models/ChartModel/index.js]: [05e9f09](https://github.com/AMP-SCZ/dpdash/pull/530/commits/05e9f090fc738b3090970d1e415c0b1966628835)

File: views/models/ChartModel/index.js (hunk 4 above)
Comment (Elkrival) [reply views/models/ChartModel/index.js]: [05e9f09](https://github.com/AMP-SCZ/dpdash/pull/530/commits/05e9f090fc738b3090970d1e415c0b1966628835)

File: views/models/ChartModel/index.js (hunk 3 above)
Comment (Elkrival) [reply views/models/ChartModel/index.js]: [05e9f09](https://github.com/AMP-SCZ/dpdash/pull/530/commits/05e9f090fc738b3090970d1e415c0b1966628835)

File: views/models/ChartModel/index.js (hunk 2 above)
Comment (Elkrival) [reply views/models/ChartModel/index.js]: [05e9f09](https://github.com/AMP-SCZ/dpdash/pull/530/commits/05e9f090fc738b3090970d1e415c0b1966628835)

File: views/models/UrlModel/index.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+const UrlModel = {
+  sanitizeUrl: (url) => {
+    url = String(url)
+
+    if (url === 'null' || url.length === 0 || url === 'about:blank')
+      return 'data:,Unsafe%20Image'
+
+    if (url.match(SAFE_URL_PATTERN) || url.match(DATA_URL_PATTERN)) return url
+
+    return `unsafe:${url}`
```
Comment (JackVCurtis) [inline views/models/UrlModel/index.js]: Can we return `data:,Unsafe%20Image` here as well?

**Review by JackVCurtis — APPROVED** [review]
One small change but looks great, I'm feeling much better about this feature

**Pull request merged**